diff --git a/src/classes/worker.ts b/src/classes/worker.ts
index b79ddf464..41835e812 100644
--- a/src/classes/worker.ts
+++ b/src/classes/worker.ts
@@ -702,7 +702,11 @@ will never work with more accuracy than 1ms. */

           // Markers should only be used for un-blocking, so we will handle them in this
           // function only.
+          const gettime = () => '[' + new Date().toJSON() + ']';
+          const id = Math.random().toString(16).slice(2);
+          console.log(`${gettime()} Pre bzpopmin ${id}`);
           const result = await bclient.bzpopmin(this.keys.marker, blockTimeout);
+          console.log(`${gettime()} Post bzpopmin ${id}`);

           if (result) {
             const [_key, member, score] = result;
diff --git a/tests/test_bulk.ts b/tests/test_bulk.ts
index cdaa97ea7..ab20c5b7c 100644
--- a/tests/test_bulk.ts
+++ b/tests/test_bulk.ts
@@ -123,18 +123,23 @@ describe('bulk jobs', () => {
     const numJobs = 6;
     const queueEvents = new QueueEvents(queueName, { connection, prefix });
     await queueEvents.waitUntilReady();
+    const gettime = () => '[' + new Date().toJSON() + ']';

     const worker = new Worker(
       queueName,
       async () => {
+        console.log(`${gettime()} Pre wait 1`);
         await delay(900);
+        console.log(`${gettime()} Post wait 1`);
       },
       { connection, prefix },
     );
     const worker2 = new Worker(
       queueName,
       async () => {
+        console.log(`${gettime()} Pre wait 2`);
         await delay(900);
+        console.log(`${gettime()} Post wait 2`);
       },
       { connection, prefix },
     );
@@ -158,10 +163,15 @@ describe('bulk jobs', () => {

     await queue.addBulk(jobs);

+    console.log(`${gettime()} pre completed`);
     await completed;
+    console.log(`${gettime()} pre close 1`);
     await worker.close();
+    console.log(`${gettime()} pre close 2`);
     await worker2.close();
+    console.log(`${gettime()} pre close queue events`);
     await queueEvents.close();
+    console.log(`${gettime()} done`);
   });

   it('should process jobs with custom ids', async () => {
