diff --git a/tests/test_bulk.ts b/tests/test_bulk.ts
index cdaa97ea7..ccf5e9d9c 100644
--- a/tests/test_bulk.ts
+++ b/tests/test_bulk.ts
@@ -123,18 +123,23 @@ describe('bulk jobs', () => {
     const numJobs = 6;
     const queueEvents = new QueueEvents(queueName, { connection, prefix });
     await queueEvents.waitUntilReady();
+    const gettime = () => '[' + new Date().toJSON() + ']';

     const worker = new Worker(
       queueName,
       async () => {
+        console.log(`${gettime()} Pre wait 1`);
         await delay(900);
+        console.log(`${gettime()} Post wait 1`);
       },
       { connection, prefix },
     );
     const worker2 = new Worker(
       queueName,
       async () => {
+        console.log(`${gettime()} Pre wait 2`);
         await delay(900);
+        console.log(`${gettime()} Post wait 2`);
       },
       { connection, prefix },
     );
